### PR TITLE
Api definition

### PR DESCRIPTION
**Pull request by majsa1**
Far from ready yet, but could use some feedback!
Not sure how much logic we will place in the backend. For example, post to user/{userId}/challenge now assumes that all the necessary tables are created in the backend based on the information sent (challenge, user_challenges and challenge_days). Also some get-requests return information from two tables.

You can upload the yaml-file to https://editor.swagger.io.

Thanks!

### CODE REVIEW

File: documentation/wasted_openAPI_specification.yaml [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+        - Users
+      summary: Modifies an existing user based on ID.
+      requestBody:
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/User'
+      responses:
+        '204':
```
Comment (ahmed-ben10) [inline documentation/wasted_openAPI_specification.yaml:94]: The status code we have now setup for editing a user is 200

File: documentation/wasted_openAPI_specification.yaml [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+        - Users
+      summary: Creates a new user.
+      requestBody:
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/User'
+      responses:
+        '201':
```
Comment (ahmed-ben10) [inline documentation/wasted_openAPI_specification.yaml:41]: The status code we have now setup for creating a user is 200

File: documentation/wasted_openAPI_specification.yaml [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+
+components:
+  securitySchemes:
+    bearerAuth:           
+      type: http
+      scheme: bearer
+      bearerFormat: JWT   
+  schemas:
+    User:
+      properties:
```
Comment (ahmed-ben10) [inline documentation/wasted_openAPI_specification.yaml:259]: is_admin property is missing

File: documentation/wasted_openAPI_specification.yaml [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+        '500':
+          $ref: '#/components/responses/ServerError'
+
+components:
+  securitySchemes:
+    bearerAuth:           
+      type: http
+      scheme: bearer
+      bearerFormat: JWT   
+  schemas:
```
Comment (ahmed-ben10) [inline documentation/wasted_openAPI_specification.yaml:257]: Do these properties represent the attributes in the entity?

**Review by ahmed-ben10 — CHANGES_REQUESTED** [review]

File: documentation/wasted_openAPI_specification.yaml (hunk 2 above)
Comment (majsa1) [reply documentation/wasted_openAPI_specification.yaml:41]: Changed both to 200

File: documentation/wasted_openAPI_specification.yaml (hunk 3 above)
Comment (majsa1) [reply documentation/wasted_openAPI_specification.yaml:259]: Does the API need to return this information (do we need it in the front end)? Or would it be for future implementations?

File: documentation/wasted_openAPI_specification.yaml (hunk 4 above)
Comment (majsa1) [reply documentation/wasted_openAPI_specification.yaml:257]: No, this is Swagger's way to indicate the authorisation scheme (you could potentially enter the token and test all paths from SwaggerUI)

Comment (majsa1) [reply documentation/wasted_openAPI_specification.yaml:257]: Do you mean schemas? Yes, those are the entities with their attributes
Securityscheme is the authorisation mechanism (you could potentially use to test the API in SwaggerUI)

**Comment by majsa1** [comment]
What do you think is better:
/challenge/{challengeId}/days
or
/userchallenge/{userChallengeId}/days?
The first option results in less calls, but the second in less complexity.

File: documentation/wasted_openAPI_specification.yaml (hunk 3 above)
Comment (ahmed-ben10) [reply documentation/wasted_openAPI_specification.yaml:360]: I think you're right. We can leave it out

**Review by ahmed-ben10 — APPROVED** [review]

**Review by ahmed-ben10 — APPROVED** [review]

**Comment by ahmed-ben10** [comment]
I would choose the /challenge variant

**Comment by majsa1** [comment]
Added groups, badges, suggestions etc, also included minor changes.

**Pull request closed without merging**